tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,460 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the in-memory filesystem class
+ */
+
+#include <sstream>
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/mem_filesystem.h"
+
+using namespace tiledb::common;
+using namespace std;
+
+namespace tiledb {
+namespace sm {
+
+class MemFilesystem::FSNode {
+ public:
+  FSNode(const string& name, FSNode* parent)
+      : name_(name)
+      , parent_(parent){};
+
+  virtual ~FSNode() = default;
+
+  /* Indicates if it's a file or a directory */
+  virtual bool is_dir() const = 0;
+
+  /* Lists the contents of a node */
+  virtual vector<string> ls(const string& full_path) const = 0;
+
+  /* Indicates if a given node is a child of this node */
+  virtual bool has_child(const string& child) const = 0;
+
+  /* Gets the size in bytes of this node */
+  virtual Status get_size(uint64_t* size) const = 0;
+
+  /* Outputs the contents of the node to a buffer */
+  virtual Status read(
+      const uint64_t offset, void* buffer, const uint64_t nbytes) const = 0;
+
+  /* Outputs the contents of a buffer to this node */
+  virtual Status append(const void* data, const uint64_t nbytes) = 0;
+
+  /* A hashtable of all the next-level subnodes of this node*/
+  unordered_map<string, unique_ptr<FSNode>> children_;
+
+  /* The name of this node */
+  string name_;
+
+  /* A pointer to the parent node of this node */
+  FSNode* parent_;
+};
+
+class MemFilesystem::File : public MemFilesystem::FSNode {
+ public:
+  /** Constructor. */
+  File(const string& name, FSNode* parent)
+      : FSNode(name, parent)
+      , data_(nullptr)
+      , size_(0){};
+
+  /** Destructor. */
+  ~File() {
+    if (data_) {
+      free(data_);
+    }
+  };
+
+  bool is_dir() const override {
+    return false;
+  }
+
+  /* Returns the full path to this file */
+  vector<string> ls(const string& full_path) const override {
+    return {full_path + name_};
+  }
+
+  bool has_child(const string&) const override {
+    return false;
+  }
+
+  Status get_size(uint64_t* size) const override {
+    *size = size_;
+    return Status::Ok();
+  }
+
+  /* Writes nbytes of data in the end of this file */
+  Status append(const void* data, const uint64_t nbytes) override {
```

Comment:
This seems to be reusing functionality we implemented in `Buffer`. Does it make sense to use that instead?

Reply:
A `buffer` would definitely work here -- however, I'm in favor of keeping this as-is because we will likely use some other memory management abstraction for performance (e.g. a memory arena) that would replace uses of `Buffer`.

Reply:
Make sense, thanks.